sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1013 @@
+# coding: utf-8
+"""
+Large Margin Nearest Neighbor Classification
+"""
+
+# Author: John Chiotellis <johnyc.code@gmail.com>
+# License: BSD 3 clause
+
+from __future__ import print_function
+from warnings import warn
+
+import numpy as np
+import time
+import sys
+from scipy.optimize import fmin_l_bfgs_b
+from scipy.sparse import csr_matrix, csc_matrix, spdiags
+
+from ..base import BaseEstimator, TransformerMixin
+from ..neighbors import NearestNeighbors
+from ..decomposition import PCA
+from ..utils import gen_batches
+from ..utils.extmath import row_norms, safe_sparse_dot
+from ..utils.random import check_random_state
+from ..utils.multiclass import check_classification_targets
+from ..utils.validation import check_is_fitted, check_array, check_X_y, \
+    check_consistent_length
+
+
+class LargeMarginNearestNeighbor(BaseEstimator, TransformerMixin):
+    """Distance Metric Learning for Large Margin Classification.
+
+    Parameters
+    ----------
+    n_features_out : int, optional (default=None)
+        Preferred dimensionality of the embedding.
+        If None it is inferred from ``init``.
+
+    init : string or numpy array, optional (default='pca')
+        Initialization of linear transformation. Possible options are 'pca',
+        'identity' and a numpy array of shape (n_features_out, n_features).
+
+    warm_start : bool, optional, (default=False)
+        If True and :meth:`fit` has been called before, the solution of the
+        previous call to :meth:`fit` is used as the initial linear
+        transformation (``init`` is ignored).
+
+    n_neighbors : int, optional (default=3)
+        Number of neighbors to use as target neighbors for each sample.
+
+    algorithm : {'auto', 'ball_tree', 'kd_tree', 'brute'}, optional
+        Algorithm used to compute the target neighbors:
+
+        - 'ball_tree' will use :class:`BallTree`
+        - 'kd_tree' will use :class:`KDTree`
+        - 'brute' will use a brute-force search.
+        - 'auto' will attempt to decide the most appropriate algorithm
+          based on the values passed to :meth:`fit` method.
+
+        Note: fitting on sparse input will override the setting of
+        this parameter, using brute force.
+
+    targets : array, shape (n_samples, n_neighbors), optional (default=None)
+        The target neighbors of each sample, namely the `n_neighbors` nearest
+        neighbors that belong to the same class. If None (default), the target
+        neighbors will be computed on the fly using the euclidean metric.
```

Comment:
why do you have an array that is data dependent in the init? is this really necessary? If so at least the docstring should explain why.

Reply:
hum but my concern is that by doing this you break the API logic. cloning estimator and splitting train and test for cross-valdiation becomes broken. I think we should this possibility out for the first version of this estimator.

Reply:
Okay.

Reply:
one way to keep this option open is to expose lmnn as a plain function (not an object estimator). Then there is no API issue

X_transformed, metric = lmmn(X, ..., targets=None, ...)

then estimator just uses the plain function in the fit